src/mcl_3dl.cpp

```diff
@@ -1182,6 +1167,50 @@ class MCL3dlNode
     pub_status_.publish(status_);
   }
 
+  void loadMapCloud(const pcl::PointCloud<PointType>::Ptr& map_cloud)
+  {
+    pc_map_.reset(new pcl::PointCloud<PointType>);
+    pc_map2_.reset();
+    pc_update_.reset();
+    pcl::VoxelGrid18<PointType> ds;
+    ds.setInputCloud(map_cloud);
+    ds.setLeafSize(params_.map_downsample_x_, params_.map_downsample_y_, params_.map_downsample_z_);
+    ds.filter(*pc_map_);
+    pc_all_accum_.reset(new pcl::PointCloud<PointType>);
+    has_map_ = true;
+
+    accumClear();
+    accum_->reset();
+
+    ROS_INFO("map original: %d points", static_cast<int>(map_cloud->points.size()));
+    ROS_INFO("map reduced: %d points", static_cast<int>(pc_map_->points.size()));
+
+    // output map for visualization purposes:
+    cbMapUpdateTimer(ros::TimerEvent());
+  }
+
+  bool cbLoadPCD(mcl_3dl_msgs::LoadPCD::Request& req, mcl_3dl_msgs::LoadPCD::Response& resp)
+  {
+    ROS_INFO("map received");
+
+    pcl::PointCloud<PointType>::Ptr pc_tmp(new pcl::PointCloud<PointType>);
+    if (pcl::io::loadPCDFile<PointType> (req.pcd_path, *pc_tmp) == -1)
+    {
+      ROS_ERROR_STREAM("Couldn't read file " << req.pcd_path);
+      has_map_ = false;
+      resp.success = false;
+      return true;
+    }
+
+    pcl_conversions::toPCL(ros::Time::now(), pc_tmp->header.stamp);
+    pc_tmp->header.frame_id = req.frame_id;
```

Comment:
`frame_id` of `pc_map_` should be `params_.frame_ids_["map"]`.
Possible options are
- transform the pcd data to `params_.frame_ids_["map"]` if `req.frame_id` is not equal to `params_.frame_ids_["map"]`
- remove `frame_id` from `mcl_3dl_msgs::LoadPCD::Request` and assume that the frame of the pcd file is `params_.frame_ids_["map"]`

Reply:
Nice one. I changed it and removed `frame_id` from the request.